OrbitGl/FrameTrack.cpp

```diff
@@ -0,0 +1,228 @@
+  std::string function_name = FunctionUtils::GetDisplayName(function_);
+  std::string name = absl::StrFormat("Frame track based on %s", function_name);
+  SetName(name);
+  SetLabel(name);
+
+  // Delta tracks are collapsed by default.
+  collapse_toggle_->SetState(TriangleToggle::State::kCollapsed,
+                             TriangleToggle::InitialStateUpdate::kReplaceInitialState);
+}
+
+float FrameTrack::GetHeaderHeight() const { return 0.f; }
+
+float FrameTrack::GetHeight() const {
+  TimeGraphLayout& layout = time_graph_->GetLayout();
+  return GetMaximumBoxHeight() + layout.GetTrackBottomMargin();
+}
+
+float FrameTrack::GetYFromDepth(uint32_t /*depth*/) const {
+  return pos_[1] - GetMaximumBoxHeight();
+}
+
+float FrameTrack::GetTextBoxHeight(const TimerInfo& timer_info) const {
+  uint64_t timer_duration_ns = timer_info.end() - timer_info.start();
+  double ratio =
+      static_cast<double>(timer_duration_ns) / static_cast<double>(stats_.average_time_ns());
+  ratio = std::min(ratio, kHeightCapAverageMultipleDouble);
+  return static_cast<float>(ratio) * GetAverageBoxHeight();
+  ;
+}
+
+Color FrameTrack::GetTimerColor(const orbit_client_protos::TimerInfo& timer_info,
+                                bool /*is_selected*/) const {
+  Vec4 min_color(76.f, 175.f, 80.f, 255.f);
+  Vec4 max_color(63.f, 81.f, 181.f, 255.f);
+  Vec4 warn_color(244.f, 67.f, 54.f, 255.f);
+
+  Vec4 color;
+  uint64_t timer_duration_ns = timer_info.end() - timer_info.start();
+
+  if (timer_duration_ns >= kHeightCapAverageMultipleUint64 * stats_.average_time_ns()) {
+    color = warn_color;
+  } else {
+    uint64_t lower_bound = stats_.min_ns();
+    uint64_t upper_bound =
+        std::min(kHeightCapAverageMultipleUint64 * stats_.average_time_ns(), stats_.max_ns());
+    timer_duration_ns = std::min(timer_duration_ns, upper_bound);
+    float fraction = static_cast<float>(timer_duration_ns - lower_bound) /
+                     static_cast<float>(upper_bound - lower_bound);
+    color = fraction * max_color + (1.f - fraction) * min_color;
+  }
+
+  if (timer_info.depth() == 0) {
+    return Color(static_cast<uint8_t>(color[0]), static_cast<uint8_t>(color[1]),
+                 static_cast<uint8_t>(color[2]), static_cast<uint8_t>(color[3]));
+  } else {
+    color = 0.8f * color;
+    return Color(static_cast<uint8_t>(color[0]), static_cast<uint8_t>(color[1]),
+                 static_cast<uint8_t>(color[2]), static_cast<uint8_t>(color[3]));
+  }
+}
+
+void FrameTrack::OnTimer(const TimerInfo& timer_info) {
+  uint64_t duration_ns = timer_info.end() - timer_info.start();
+  stats_.set_count(stats_.count() + 1);
+  stats_.set_total_time_ns(stats_.total_time_ns() + duration_ns);
+  stats_.set_average_time_ns(stats_.total_time_ns() / stats_.count());
+
+  if (duration_ns > stats_.max_ns()) {
+    stats_.set_max_ns(duration_ns);
+  }
+  if (stats_.min_ns() == 0 || duration_ns < stats_.min_ns()) {
```

Comment:
This assumes that a duration is always >= 1ns, which I think is the case in practice, but maybe we should put in a check to be sure? Or possibly initialize stats_min_ns() to std::numeric_limits<uint64_t>::max()?

Reply:
But doesn't it still work correctly, though? If duration_ns is 0 for at least one timer, then we'll just set the min to 0 again in the if body.

Reply:
Yes you're right, nevermind :-)